### PR TITLE
SE2933: Adding new notebook: Requesting logs of inline valuation oper…

### PR DESCRIPTION
**Pull request by armlowe**
…ations using the insights API

# Pull Request Checklist

- [x] Changes follow the [style guide](https://github.com/finbourne/sample-notebooks/blob/master/docs/FINBOURNE%20notebook%20style%20guide.ipynb)
- [x] Tests pass
- [x] Raised the PR against the `develop` branch

# Description of the PR

Adding a notebook to demonstrate how to query a set of insights requests using the Insights API and represent the data in a pandas data frame.

### CODE REVIEW

File: examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb [hunk 1]
```diff
@@ -0,0 +1,857 @@
+{
```
Comment (rizwansaeed) [inline examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:1]: <p>missing the meta data as per the template</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/615/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (rizwansaeed) [inline examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #54.    i_api_factory = ApiClientFactorySdk(
</pre><p> </p><p>not a huge fan of the name <code>ApiClientFactorySdk</code>, maybe namespace the import as you've done with lusid   </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/615/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (rizwansaeed) [inline examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #69.    valuation_date = datetime(year=2023, month=1, day=26, tzinfo=pytz.UTC)
</pre><p> </p><p>i think you will need to make this a str using <code>.isoformat()</code>,the effectiveAt is actually a string as it can be a cut label. there is a recent change gone with tightens up the validation, try running your nb in fbn-ci to confirm  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/615/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (rizwansaeed) [inline examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #70.    portfolio_code = "Insights_inline_valuation_test"
</pre><p> </p><p>this looks like its unused</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/615/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (rizwansaeed) [inline examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:1]: <p>fyi, we will shortly have <code>FBN_LUSID_INSIGHTS_URL</code> as an env var </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/615/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (rizwansaeed) [inline examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.    recipe_scope="Insights_Inline_valuation_NB"
</pre><p> </p><p>would it make sense for <code>global_scope</code> to have this value too? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/615/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (rizwansaeed) [inline examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:1]: <p>Add some commentary to explain you are using the Insights api to query the logs</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/615/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

Comment (rizwansaeed) [inline examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    #In busy systems it can take a few seconds for the insights logs to be available, this function will
</pre><p> </p><p>Can we reword this to something more like:</p><p>"Logs are written asynchronously and there may be a lag for them to be queryable after writing. This function..."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/615/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='30'/>

File: examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb [hunk 2]
```diff
@@ -0,0 +1,857 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d93eef1e",
+   "metadata": {},
+   "source": [
+    "## 1. Setup LUSID and Insights\n",
+    "\n",
+    "This section will load the relevant python packages as well as create the LUSID and Insights API clients. Th LUSID API clients will be used to create the valuation whereas the Insights API client will be used to request the logs and request/responses."
```
Comment (liberty-askew) [inline examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:26]: Minor typo "Th"

File: examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb [hunk 3]
```diff
@@ -0,0 +1,857 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "id": "0e48fe19",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def spot_request(from_ccy, to_ccy, rate, valuation_date):\n",
+    "            return models.UpsertQuoteRequest(\n",
```
Comment (liberty-askew) [inline examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:199]: Over indented here

File: examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb [hunk 4]
```diff
@@ -0,0 +1,857 @@
...
+   "source": [
+    "## 3. Define Weighted Instruments"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "7dc53f03",
+   "metadata": {},
+   "source": [
+    "Define 5 weighted equity instruments with the same 'ClientInternal' identifiers as those of the 5 quotes, a quantity of 1 and a currency of GBP (same as for the quotes). Put these 5 definitions into a list to be passed into the successful inline valuation request in Section 4."
```
Comment (liberty-askew) [inline examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:238]: Should this be section 5 not 4?

File: examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb [hunk 5]
```diff
@@ -0,0 +1,857 @@
...
+    "recipe_code=\"Insights_Inline_valuation_NB\"\n",
+    "\n",
+    "\n",
+    "# Create a recipe to perform a valuation\n",
+    "configuration_recipe = models.ConfigurationRecipe(\n",
+    "    scope=recipe_scope,\n",
+    "    code=recipe_code,\n",
+    "    market=models.MarketContext(\n",
+    "        market_rules=[\n",
+    "            # define how to resolve the quotes\n",
```
Comment (liberty-askew) [inline examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:301]: Capital D

File: examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb [hunk 6]
```diff
@@ -0,0 +1,857 @@
...
+       "4              GBP  "
+      ]
+     },
+     "execution_count": 8,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "#Run inline valuation in JPY\n",
```
Comment (liberty-askew) [inline examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:524]: Space between # R

File: examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb [hunk 7]
```diff
@@ -0,0 +1,857 @@
...
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "ERROR : One or more of the bits of input data provided were not valid\n"
+     ]
+    }
+   ],
+   "source": [
+    "#Run inline Valuation for an invalid currency code\n",
```
Comment (liberty-askew) [inline examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:567]: Space between # R

File: examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb [hunk 8]
```diff
@@ -0,0 +1,857 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "id": "a3268f11",
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "#In busy systems it can take a few seconds for the insights logs to be available, this function will\n",
```
Comment (liberty-askew) [inline examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:611]: Space here

File: examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb [hunk 9]
```diff
@@ -0,0 +1,857 @@
...
+    "import backoff\n",
+    "\n",
+    "# Set pandas dataframe display formatting\n",
+    "pd.set_option('display.max_columns', None)\n",
+    "pd.options.display.float_format = '{:,.2f}'.format\n",
+    "\n",
+    "# Get secrets path for API client\n",
+    "secrets_path = os.getenv(\"FBN_SECRETS_PATH\")\n",
+    "\n",
+    "#set a correlation ID in order to find all logs related to this session\n",
```
Comment (liberty-askew) [inline examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:69]: Space and capital S

File: examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb [hunk 10]
```diff
@@ -0,0 +1,857 @@
...
+   "source": [
+    "# Requesting details of inline valuation operations using the insights API\n",
+    "\n",
+    "This notebook demonstrates the basic usage of the insights API to request logs and the detailed request/response bodies for a given operation. In this notebook, we will perform the following workflow:\n",
+    "\n",
+    "- We run a valuation on an inline portfolio in GBP for a JPY/GBP FX rates which succeeds\n",
+    "- In the same portfolio, we attempt to run another valuation for an AUD/GBP FX rate which fails due to a badly constructed request\n",
+    "- We then use the Insights Requests API to retreive all Weighted instrument valuation requests from the previous 5 minutes and display them along with any errors in a data frame\n",
+    "\n",
+    "The reason that an inline valuation is used in this instance is to make the notebook easier to re-run and that meaningful data can be extracted from the response body to indicate the success or failure of the operation"
```
Comment (liberty-askew) [inline examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:16]: fullstop here

File: examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb [hunk 11]
```diff
@@ -0,0 +1,857 @@
...
+   "source": [
+    "### 6.1 Get the request logs"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "12fae8ea",
+   "metadata": {},
+   "source": [
+    "Now that we have a successful and unsuccesful request in the system, we can request the logs, which will give us an overview of each operation. Here we make a request for all Weighted instrument valuations made in the last 5 minutes. We also provide the correlation ID which was set in section 1 to filter for only transactions from this session"
```
Comment (liberty-askew) [inline examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:599]: Fullstop here

File: examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb [hunk 12]
```diff
@@ -0,0 +1,905 @@
+{
```
Comment (Tomwp) [inline examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:1]: <p>Can you make mention of the fact that you're using a correlation ID. Also if you can reference this KB article: https://support.lusid.com/knowledgebase/article/KA-01714/en-us.</p><p> </p><p>Also the very last sentence should probably read: "request/response bodies".</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/615/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (Tomwp) [inline examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:1]: <p>typo - 'disaplays'</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/615/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='33'/>

Comment (Tomwp) [inline examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:1]: <p>Maybe reword - 'If we want to examine the errors further, we can...'. You use the word detail twice in the same sentence.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/615/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='35'/>

Comment (Tomwp) [reply examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:1]: <p>In section 1. where you talk about the setup. </p>

File: examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb [hunk 13]
```diff
@@ -0,0 +1,680 @@
...
```
Comment (Tomwp) [inline examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:1]: <p>Two things:</p><ol><li>Capitalise Insights. I would also re-write it: "Requesting log details of an inline valuation using the Insights API". Also update this in the toggle code section.</li><li>See below:</li></ol><p> </p><p>I think we should re-write the first two bullets as:</p><p> </p><ul><li>We run a valuation of an inline GBP portfolio with a reporting currency of JPY whereby we provide the requisite JPY/GBP FX rate to ensure our valuation runs successfully.</li><li>In the same portfolio, we attempt to run the valuation with a reporting currency of AUD but where we don't provide the requisite AUD/GBP rate. This valuation will fail.</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/615/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (Tomwp) [inline examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #65.    
</pre><p>Remove extra whitespace</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/615/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (Tomwp) [inline examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #14.                    
</pre><p>Remove extra whitespace</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/615/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (Tomwp) [inline examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #15.            
</pre><p>Remove extra whitespace</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/615/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (Tomwp) [inline examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #7.    
</pre><p>Remove extra whitespace</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/615/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Comment (Tomwp) [inline examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #16.    
</pre><p>Remove extra whitespace</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/615/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

Comment (Tomwp) [inline examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #25.    
</pre><p>Remove extra whitespace</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/615/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='34'/>

Comment (Tomwp) [reply examples/use-cases/audit-trail/Requesting logs of inline valuation operations using the insights API.ipynb:1]: Makes sense - that wording looks good.

File: examples/use-cases/audit-trail/Requesting log details using the Insights API.ipynb [hunk 14]
```diff
@@ -0,0 +1,673 @@
+{
```
Comment (Tomwp) [inline examples/use-cases/audit-trail/Requesting log details using the Insights API.ipynb:1]: <p>Weighted instruments should be lower case</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/615/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (Tomwp) [inline examples/use-cases/audit-trail/Requesting log details using the Insights API.ipynb:1]: <p>Capitalize Python.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/615/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (Tomwp) [inline examples/use-cases/audit-trail/Requesting log details using the Insights API.ipynb:1]: <p>Weighted should be lower case</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/615/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='30'/>

**Review by Tomwp — APPROVED** [review]
Approved
